### PR TITLE
Prefilters

### PR DESCRIPTION
**Pull request by d-callan**
prefilter features/ taxa/ pathway/ etc in AbundanceData based on predicates. ive defined some potentially handy predicate functions as well, but for now am only using the one that removes features if fewer than 50% of samples have data.

implemented for the correlation method, but could be useful for others as well maybe

### CODE REVIEW

File: R/method-correlation.R [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## prefilter predicates
+halfOfSamplesHaveNonZeroAbundance <- function(x){sum(x > 0) >= length(x)/2}
```
Comment (asizemore) [inline R/method-correlation.R:2]: this predicate is more general than "samples" and "abundances". Can we generalize the name to something like "halfAreNonZero"?

File: R/method-correlation.R [hunk 2]
```diff
@@ -163,6 +170,9 @@ buildCorrelationComputeResult <- function(corrResult, data1, data2 = NULL, metho
 #' @return a ComputeResult object
 #' @export
 setMethod("correlation", signature("AbundanceData", "missing"), function(data1, data2, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+  #prefilters applied
+  data1 <- pruneFeatures(data1, halfOfSamplesHaveNonZeroAbundance, verbose)
```
Comment (asizemore) [inline R/method-correlation.R:174]: If we think we're going to let the user make some of these pruning decisions in the future, why not add arguments to our correlation functions and just set these as defaults?

File: R/methods-AbundanceData.R [hunk 3]
```diff
@@ -113,5 +113,35 @@ setMethod("removeIncompleteSamples", signature("AbundanceData"), function(object
...
+#' @return AbundanceData with modified data slot
+#' @export
+setGeneric("pruneFeatures",
+  function(object, predicate, verbose = c(TRUE, FALSE)) standardGeneric("pruneFeatures"),
+  signature = c("object")
+)
+
+#'@export 
+setMethod("pruneFeatures", signature("AbundanceData"), function(object, predicate, verbose = c(TRUE, FALSE)) {
+  df <- getAbundances(object)
```
Comment (asizemore) [inline R/methods-AbundanceData.R:137]: + matchArgs for verbose

File: tests/testthat/test-correlation.R [hunk 4]
```diff
@@ -108,9 +108,9 @@ test_that('correlation returns an appropriately structured result for abundance
   statsData <- result@statistics@statistics
   expect_s3_class(statsData, 'data.frame')
   expect_equal(names(statsData), c('data1','data2','correlationCoef','pValue'))
-  expect_equal(nrow(statsData), (ncol(testOTU) - 1) * length(veupathUtils::findNumericCols(sampleMetadata@data))) # Should be number of taxa * number of metadata vars
-  expect_equal(as.character(unique(statsData$data1)), names(testOTU)[2:length(names(testOTU))])
-  expect_equal(as.character(unique(statsData$data2)), c('entity.contA', 'entity.contB', 'entity.contC'))
+  expect_equal(nrow(statsData), 36) # Should be number of taxa * number of metadata vars, less pruned taxa
+  #expect_equal(as.character(unique(statsData$data1)), names(testOTU)[2:length(names(testOTU))])
+  #expect_equal(as.character(unique(statsData$data2)), c('entity.contA', 'entity.contB', 'entity.contC'))
```
Comment (asizemore) [inline tests/testthat/test-correlation.R:113]: this one should still work, no?

**Review by asizemore — COMMENTED** [review]
Mostly questions. I also think it'd be great to replace the hardcoded numbers in the tests with some logic just in case our test data changes

File: R/method-correlation.R (hunk 2 above)
Comment (d-callan) [reply R/method-correlation.R:174]: it probably makes sense.. i expect were going to iterate here though, and if im going to be revisiting this anyway id rather make api changes once i know more about our needs.

File: tests/testthat/test-correlation.R (hunk 4 above)
Comment (d-callan) [reply tests/testthat/test-correlation.R:113]: why?

File: R/method-correlation.R (hunk 2 above)
Comment (d-callan) [reply R/method-correlation.R]: ... i guess the one about proportion of samples w non-zero values will probably def be around. maybe i can work that one in at least.

Comment (d-callan) [reply R/method-correlation.R]: ive had a better idea.. hold please

**Comment by d-callan** [comment]
> I also think it'd be great to replace the hardcoded numbers in the tests with some logic just in case our test data changes

maybe im being dense.. but i didnt see a good way to do this. how many taxa get removed depends pretty directly on the test data used, and i didnt see a good way to find the number without using the same functions i was testing, which kind of defeats the purpose.

**Comment by d-callan** [comment]
I've realized belatedly that we could pass the default filters/ filters used in the response for clarity. I can do that separately once I get things working.

**Pull request merged**